core/src/main/scala/org/ensime/core/javac/JavaCompiler.scala

```diff
@@ -0,0 +1,178 @@
+    pathToPoint(file, offset) flatMap {
+      case (info: CompilationInfo, path: TreePath) =>
+        docSignature(info, path)
+    }
+  }
+
+  def askCompletionsAtPoint(
+    file: SourceFileInfo, offset: Int, maxResults: Int, caseSens: Boolean
+  ): Future[CompletionInfoList] = Future {
+    completionsAt(file, offset, maxResults, caseSens)
+  }
+
+  protected def pathToPoint(file: SourceFileInfo, offset: Int): Option[(CompilationInfo, TreePath)] = {
+    val infos = typecheckForUnits(List(getJavaFileObject(file)))
+    infos.headOption.flatMap { info =>
+      val path = Option(new TreeUtilities(info).pathFor(offset))
+      path.map { p => (info, p) }
+    }
+  }
+
+  protected def scopeForPoint(file: SourceFileInfo, offset: Int): Option[(CompilationInfo, Scope)] = {
+    val infos = typecheckForUnits(List(getJavaFileObject(file)))
+    infos.headOption.flatMap { info =>
+      val path = Option(new TreeUtilities(info).scopeFor(offset))
+      path.map { p => (info, p) }
+    }
+  }
+
+  private def typeMirrorToTypeInfo(tm: TypeMirror): TypeInfo = {
+    BasicTypeInfo(tm.toString, -1, DeclaredAs.Class, tm.toString, List(), List(), Some(EmptySourcePosition()), None)
+  }
+
+  private def getTypeMirror(info: CompilationInfo, offset: Int): Option[TypeMirror] = {
+    val path = Option(new TreeUtilities(info).pathFor(offset))
+    // Uncomment to debug the AST path.
+    //for (p <- path) { for (t <- p) { System.err.println(t.toString()) } }
+    path.flatMap { p => Option(info.getTrees().getTypeMirror(p)) }
+  }
+
+  private def typecheckAll(files: Iterable[JavaFileObject]): Unit = {
+    val task = compiler.getTask(null, fileManager, listener, List(
+      "-cp", cp, "-Xlint:all", "-proc:none"
+    ), null, asJavaIterable(files)).asInstanceOf[JavacTask]
+    val t = System.currentTimeMillis()
+    task.parse()
+    task.analyze()
+    System.err.println("Parsed and analyzed: " + (System.currentTimeMillis() - t) + "ms")
+  }
+
+  private def typecheckForUnits(files: Iterable[JavaFileObject]): Vector[CompilationInfo] = {
+    val task = compiler.getTask(null, fileManager, silencer, List(
+      "-cp", cp, "-Xlint:none", "-proc:none"
+    ), null, asJavaIterable(files)).asInstanceOf[JavacTask]
+    val t = System.currentTimeMillis()
+    val units = task.parse().map(new CompilationInfo(task, _)).toVector
+    task.analyze()
+    System.err.println("Parsed and analyzed for trees: " + (System.currentTimeMillis() - t) + "ms")
+    units
+  }
+
+  private class JavaObjectWithContents(val f: File, val contents: String)
+      extends SimpleJavaFileObject(f.toURI, JavaFileObject.Kind.SOURCE) {
+    override def getCharContent(ignoreEncodingErrors: Boolean): CharSequence = contents
+  }
+
+  private class JavaObjectFromFile(val f: File)
+      extends SimpleJavaFileObject(f.toURI, JavaFileObject.Kind.SOURCE) {
+    override def getCharContent(ignoreEncodingErrors: Boolean): CharSequence = {
+      FileUtils.readFile(f, charset) match {
```

Comment:
I think we have a cleaner util for this now by importing the `org.ensime.util.file._` syntax. I'm not a big fan of the `Either[Failure, Success]` approach, especially when its just wrapping an exception anyway (there is merit if errors are common and can be detected without an exception because they can be a lot faster).